### PR TITLE
openstack: Add missing security groups for cilium etcd

### LINKED ISSUE #15765: Cilium EtcdManaged SecurityGroups missing
/kind bug

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

1.27.0 (from release-1.27 branch)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

1.27.4

**3. What cloud provider are you using?**

openstack

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

Create a cluster using cilium networking with etcdManaged enabled.

**5. What happened after the commands executed?**

All mentioned ports need to be allowed from Master to Master via TCP.

Without 2382 etcd-cilium on the node will complain:

```
{"level":"warn","ts":"2023-08-10T18:45:31.797934Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"a24fe372aee73c54","rtt":"5.905281ms","error":"dial tcp 10.151.99.217:2382: i/o timeout"}
```

Without 3991 the cilium-agent pods will complain:

```
cilium-5z9dj cilium-agent level=error msg="Unable to initialize local node. Retrying..." error="timeout while retrieving initial list of objects from kvstore" subsys=nodediscovery
```

Without 4003 (Master to Master) the cilium-agent pods will complain

```
cilium-hgldj cilium-agent {"level":"warn","ts":"2023-08-10T18:47:39.865915Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0014141c0/api.internal.zcluster2.k8s.local:4003","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
```

Without 4003 (Workers to Master) the cilium-agent pods will complain:

```
cilium-agent level=error msg="Unable to initialize local node. Retrying..." error="timeout while retrieving initial list of objects from kvstore" subsys=nodediscovery
```

**6. What did you expect to happen?**

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

Relevant part:

```yaml
spec:
  network:
    cilium:
      etcdManaged: true
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**

I will provide a fix via PR.

### PR DESCRIPTION
**Pull request by zadjadr**
Fixes #15765

### CODE REVIEW

File: pkg/model/openstackmodel/firewall.go [hunk 1]
```diff
@@ -167,6 +167,38 @@ func (b *FirewallModelBuilder) addETCDRules(c *fi.CloudupModelBuilderContext, sg
...
+				PortRangeMax: i(wellknownports.EtcdCiliumClientPort),
+			}
+			b.addDirectionalGroupRule(c, masterSG, masterSG, etcdCiliumPeerRule)
+			b.addDirectionalGroupRule(c, masterSG, masterSG, etcdCiliumGRPCRule)
+			b.addDirectionalGroupRule(c, masterSG, masterSG, etcdCiliumClientRule)
+		}
+	}
+
 	b.addDirectionalGroupRule(c, masterSG, masterSG, etcdRule)
 	b.addDirectionalGroupRule(c, masterSG, masterSG, etcdPeerRule)
```
Comment (hakman) [inline pkg/model/openstackmodel/firewall.go:203]: Please move the block after these 2 lines.

File: pkg/model/openstackmodel/firewall.go [hunk 2]
```diff
@@ -167,6 +167,38 @@ func (b *FirewallModelBuilder) addETCDRules(c *fi.CloudupModelBuilderContext, sg
 		PortRangeMin: i(2380),
 		PortRangeMax: i(2381),
 	}
+	if b.Cluster.Spec.Networking.Cilium != nil {
+		if b.Cluster.Spec.Networking.Cilium.EtcdManaged {
```
Comment (hakman) [inline pkg/model/openstackmodel/firewall.go:171]:
```suggestion
	if b.Cluster.Spec.Networking.Cilium != nil && b.Cluster.Spec.Networking.Cilium.EtcdManaged {
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by olemarkus** [comment]
Approach and rules looks good to me.

**Comment by zadjadr** [comment]
Seems like I missing that 4003 from workers to master was also needed.
The cilium-agent pods on the workers would complain:

```
cilium-agent level=error msg="Unable to initialize local node. Retrying..." error="timeout while retrieving initial list of objects from kvstore" subsys=nodediscovery
```

**Comment by zadjadr** [comment]
Sorry, this should be the final addition.. When creating a new cluster, etcd tries to also check port 3992 (Master to Master)

**Review by hakman — APPROVED** [review]

**Pull request merged**